compiler/rustc_middle/src/query/mod.rs

```diff
@@ -1860,13 +1860,21 @@ rustc_queries! {
         eval_always
         desc { "calculating the stability index for the local crate" }
     }
-    query crates(_: ()) -> &'tcx [CrateNum] {
+    // All loaded crates, including those loaded purely for doc links or diagnostics.
+    // (Diagnostics include lints, so speculatively loaded crates may occur in successful
+    // compilation even without doc links.)
+    // Should be used when encoding crate metadata (and therefore when generating crate hash,
+    // depinfo and similar things), to avoid dangling crate references in other encoded data,
+    // like source maps.
+    // May also be used for diagnostics - if we are loading a crate anyway we can suggest some
+    // items from it as well.
+    query all_crates(_: ()) -> &'tcx [CrateNum] {
         eval_always
         desc { "fetching all foreign CrateNum instances" }
     }
     // Crates that are loaded non-speculatively (not for diagnostics or doc links).
-    // FIXME: This is currently only used for collecting lang items, but should be used instead of
-    // `crates` in most other cases too.
+    // Should be used to maintain observable language behavior, for example when collecting lang
+    // items or impls from all crates, or collecting libraries to link.
     query used_crates(_: ()) -> &'tcx [CrateNum] {
```

Comment:
"used" kind of sounds like if I add a crate as a dependency but don't use it, it doesn't show up.

Maybe `required_crates`?

Also `all_crates` could maybe need a more scary name, like `crates_including_speculative` or so?

Reply:
`crates_including_speculative` looks fine to me.
Not sure about `required_crates`.

Ideally I'd turn the `used` bool in crate loader into a 3-variant enum:
```rust
enum WhatCrateIsLoadedFor {
  ForRealThings,
  // Need to be encoded into metadata in any case
  ForDocLinksAndDiagnosticsOnly,
  // It may be possible to drop these ones from metadata, if we clean their traces from source map somehow
  ForDiagnosticsOnly, 
}
```
Maybe even merge it with existing `enum CrateDepKind` which is sort of similar.

Let's keep its pre-existing naming for now.